src/main/java/org/jenkinsci/plugins/registry/notification/events/EventType.java

```diff
@@ -0,0 +1,64 @@
+package org.jenkinsci.plugins.registry.notification.events;
+
+import hudson.EnvVars;
+import hudson.model.AbstractDescribableImpl;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.registry.notification.webhook.PushNotification;
+import org.jenkinsci.plugins.registry.notification.webhook.WebHookPayload;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import javax.annotation.Nonnull;
+import java.util.logging.Logger;
+
+public abstract class EventType extends AbstractDescribableImpl<EventType> {
+    private static final Logger logger = Logger.getLogger(EventType.class.getName());
+    public static final String KEY_DOCKER_IMAGE_DIGEST = WebHookPayload.PREFIX + "DOCKER_IMAGE_DIGEST";
+
+    public boolean accepts(@Nonnull String jsonTypeCode ){
+        return jsonTypeCode.equals(getType());
+    }
+
+    public void buildEnvironment(@Nonnull EnvVars envs, @Nonnull PushNotification push) {
+        // Each event type can have a unique time stamp field name
+        WebHookPayload payload = push.getWebHookPayload();
+        if( payload != null ){
+            JSONObject data = payload.getData();
+            if ( data != null) {
+                JSONObject contents = data.getJSONObject("contents");
+                if (contents != null) {
+                    DateTimeFormatter parser = ISODateTimeFormat.dateTimeParser();
+                    push.setPushedAt(parser.parseDateTime(getTimeStamp(contents)).toDate());
```

Comment:
Not really good idea to mutate the data here, it's a bit unexpected.

Also potential `ParseException` could be thrown here breaking the build.

Reply:
Better to have a `push.getPushedAt` instead.

Reply:
Moved this here from the DockerTrustedRegistryWebHookPayload constructor to eliminate coupling to EventType.   I can put logic back there and take the timestamp members out of the EventType classes if you think that's better.

That will, however make the JSON tokens for the different events need to be available or copied into that class.

Reply:
I'm not seeing anything that throws a ParseException here either - what am I missing?

Reply:
@rsandell - Please see above comment, not finding anything that throws that exception

Reply:
`parser.parseDateTime`

Reply:
Actually in Joda land it throws `IllegalArgumentException` instead of a `ParserException`. See http://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeFormatter.html#parseDateTime-java.lang.String-

Reply:
Ah.  Yes it does.  What is the correct behavior for a plugin to do for a Throwable like that?  Log it and leave the timestamp null or just let it bubble up?